File: sdk/npm/src/providers/webmailtemp.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'webmailtemp';
const BASE_URL = 'https://webmailtemp.com';

interface CreateResponse {
  success?: boolean;
  email?: string;
  username?: string;
  created_at?: number;
  ttl?: number;
}

interface CheckResponse {
  success?: boolean;
  emails?: Array<Record<string, unknown>>;
}

interface TokenData {
  u: string;
  c: string;
}

function encodeToken(data: TokenData): string {
  return `wmt1:${Buffer.from(JSON.stringify(data), 'utf8').toString('base64')}`;
}

function decodeToken(token: string): TokenData {
  if (!token.startsWith('wmt1:')) throw new Error('webmailtemp: invalid token');
  const raw = Buffer.from(token.slice(5), 'base64').toString('utf8');
  const data = JSON.parse(raw) as Partial<TokenData>;
  if (!data.u || !data.c) throw new Error('webmailtemp: invalid token data');
  return { u: data.u, c: data.c };
}

function flattenMessage(raw: Record<string, unknown>, recipient: string): Record<string, unknown> {
  return {
    id: raw.id,
    from: raw.from,
    to: recipient,
    subject: raw.subject,
    text: raw.body,
    html: raw.html,
    date: raw.received_at || raw.timestamp,
    attachments: raw.attachments,
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${BASE_URL}/api/create`, {
    headers: {
      Accept: 'application/json',
      'User-Agent': 'Mozilla/5.0',
    },
  });
  if (!response.ok) {
    throw new Error(`webmailtemp create http ${response.status}`);
  }
  const cookie = response.headers.get('set-cookie')?.split(';')[0] || '';
  const data = (await response.json()) as CreateResponse;
  const email = String(data.email || '').trim();
  const username = String(data.username || '').trim();
  if (!data.success || !email || !email.includes('@') || !username || !cookie) {
    throw new Error('webmailtemp: invalid create response');
  }
  return {
    channel: CHANNEL,
    email,
    token: encodeToken({ u: username, c: cookie }),
    expiresAt: typeof data.ttl === 'number' ? Date.now() + data.ttl * 1000 : undefined,
    createdAt: typeof data.created_at === 'number' ? new Date(data.created_at * 1000).toISOString() : undefined,
  };
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const address = String(email || '').trim();
  if (!address) throw new Error('webmailtemp: empty email');
  const session = decodeToken(token);

  const response = await fetchWithTimeout(`${BASE_URL}/api/check/${encodeURIComponent(session.u)}`, {
    headers: {
      Accept: 'application/json',
      Cookie: session.c,
      'User-Agent': 'Mozilla/5.0',
    },
  });
  if (!response.ok) {
    throw new Error(`webmailtemp check http ${response.status}`);
  }
  const data = (await response.json()) as CheckResponse;
  const messages = Array.isArray(data.emails) ? data.emails : [];
  return messages.map((row) => normalizeEmail(flattenMessage(row, address), address));
}

File: sdk/go/provider/webmailtemp.go
package provider

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"
	"time"

	http "github.com/bogdanfinn/fhttp"
)

const webmailtempBase = "https://webmailtemp.com"

type webmailtempToken struct {
	Username string `json:"u"`
	Cookie   string `json:"c"`
}

type webmailtempCreateResponse struct {
	Success   bool    `json:"success"`
	Email     string  `json:"email"`
	Username  string  `json:"username"`
	CreatedAt float64 `json:"created_at"`
	TTL       int64   `json:"ttl"`
}

func webmailtempEncodeToken(data webmailtempToken) (string, error) {
	raw, err := json.Marshal(data)
	if err != nil {
		return "", err
	}
	return "wmt1:" + base64.StdEncoding.EncodeToString(raw), nil
}

func webmailtempDecodeToken(token string) (webmailtempToken, error) {
	if !strings.HasPrefix(token, "wmt1:") {
		return webmailtempToken{}, fmt.Errorf("webmailtemp: invalid token")
	}
	raw, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(token, "wmt1:"))
	if err != nil {
		return webmailtempToken{}, err
	}
	var data webmailtempToken
	if err := json.Unmarshal(raw, &data); err != nil {
		return webmailtempToken{}, err
	}
	if data.Username == "" || data.Cookie == "" {
		return webmailtempToken{}, fmt.Errorf("webmailtemp: invalid token data")
	}
	return data, nil
}

func webmailtempSessionCookie(resp *http.Response) string {
	for _, value := range resp.Header.Values("Set-Cookie") {
		if p := strings.TrimSpace(strings.Split(value, ";")[0]); p != "" {
			return p
		}
	}
	return ""
}

func WebmailtempGenerate() (*CreatedMailbox, error) {
	req, err := http.NewRequest(http.MethodGet, webmailtempBase+"/api/create", nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	cookie := webmailtempSessionCookie(resp)
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("webmailtemp create: http %d", resp.StatusCode)
	}
	var data webmailtempCreateResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	if !data.Success || !strings.Contains(data.Email, "@") || data.Username == "" || cookie == "" {
		return nil, fmt.Errorf("webmailtemp: invalid create response")
	}
	token, err := webmailtempEncodeToken(webmailtempToken{Username: data.Username, Cookie: cookie})
	if err != nil {
		return nil, err
	}
	var expiresAt any
	if data.TTL > 0 {
		expiresAt = time.Now().Add(time.Duration(data.TTL) * time.Second).UnixMilli()
	}
	return &CreatedMailbox{
		Channel:   "webmailtemp",
		Email:     strings.TrimSpace(data.Email),
		Token:     token,
		ExpiresAt: expiresAt,
	}, nil
}

func webmailtempFlatten(raw map[string]any, recipient string) map[string]any {
	return map[string]any{
		"id":          raw["id"],
		"from":        raw["from"],
		"to":          recipient,
		"subject":     raw["subject"],
		"text":        raw["body"],
		"html":        raw["html"],
		"date":        raw["received_at"],
		"attachments": raw["attachments"],
	}
}

func WebmailtempGetEmails(token string, email string) ([]NormEmail, error) {
	address := strings.TrimSpace(email)
	if address == "" {
		return nil, fmt.Errorf("webmailtemp: empty email")
	}
	session, err := webmailtempDecodeToken(token)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/api/check/%s", webmailtempBase, url.PathEscape(session.Username)), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Cookie", session.Cookie)
	req.Header.Set("User-Agent", "Mozilla/5.0")
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("webmailtemp check: http %d", resp.StatusCode)
	}
	var data struct {
		Emails []map[string]any `json:"emails"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(data.Emails))
	for _, row := range data.Emails {
		out = append(out, NormalizeMap(webmailtempFlatten(row, address), address))
	}
	return out, nil
}
